marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -90,100 +183,149 @@ instance FromField C.TxIx where
+toRows event =  event ^. utxoEventUtxos & traversed %~ f
+    where
+        f :: Utxo -> UtxoRow
+        f  u = UtxoRow u (event ^. utxoEventSlotNo ) (event ^. utxoEventBlockNo)
+
+-- | only store rows in the address list.
+addressFilteredRows :: C.AddressAny -> [UtxoEvent] -> [UtxoRow]
+addressFilteredRows addr = (concatMap toRows ) . findByAddress addr
+
+-- | Query the data stored in the indexer
+-- Quries SQL + buffered data, where buffered data is the data that will be batched to SQL
 query
-  :: UtxoIndex
-  -> C.AddressAny
-  -> [UtxoUpdate]
-  -> IO Result
-query ix addr updates = do
-  -- SELECT all utxos that have not been spent.
-  let c = ix ^. Ix.handle
-  -- Create indexes initially. When created this should be a no-op.
-  SQL.execute_ c "CREATE INDEX IF NOT EXISTS utxo_address ON utxos (address)"
-  SQL.execute_ c "CREATE INDEX IF NOT EXISTS utxo_refs ON utxos (txId, inputIx)"
-  SQL.execute_ c "CREATE INDEX IF NOT EXISTS spent_refs ON spent (txId, inputIx)"
-
-  -- Perform the db query
-  storedUtxos <- SQL.query c "SELECT address, txId, inputIx FROM utxos LEFT JOIN spent ON utxos.txId = spent.txId AND utxos.inputIx = spent.inputIx WHERE utxos.txId IS NULL AND utxos.address = ?" (Only addr)
-  let memoryUtxos  = concatMap (filter (onlyAt addr) . toRows) updates
-      spentOutputs = foldMap _inputs updates
-  buffered <- Ix.getBuffer $ ix ^. Ix.storage
-  let bufferedUtxos = concatMap (filter (onlyAt addr) . toRows) buffered
-  pure . Just $ storedUtxos ++ bufferedUtxos ++ memoryUtxos
-              -- Remove utxos that have been spent (from memory db).
-              & filter (\u -> not (_reference u `Set.member` spentOutputs))
-              & map _reference
+  :: UtxoIndex                  -- ^ in-memory indexer
+  -> C.AddressAny               -- ^ Address to filter for
+  -> [UtxoEvent]                -- ^ volatile events that may be rollbacked
+  -> IO Result                  -- ^ search results
+query ix addr volatiles = do
+  diskStored <-
+      SQL.query
+        (ix ^. Ix.handle)
+        "SELECT u.address, u.txId, u.txIx, u.datum, u.datumHash, u.value, u.slotNo, u.blockNo FROM unspent_transactions u LEFT JOIN spent s ON u.txId = s.txId AND u.txIx = s.txIx WHERE u.address = ?"
+        (Only addr) :: IO[UtxoRow]
+  buffered <- Ix.getBuffer $ ix ^. Ix.storage :: IO [UtxoEvent]
+  let events = volatiles ++ buffered
+  pure . Just $
+      ( concatMap toRows . fmap rmSpentUtxos . (findByAddress addr) $ events)
+      `union`
+      diskStored
+
+-- | Query the data stored in the indexer as a whole from:
+    -- + volatile  : in-memory, datat that may rollback
+    -- + diskStore : on-disk
+    -- + buffered  : in-memeoy, data that will flush to storage
+queryPlusVolatile
+  :: UtxoIndex                  -- ^ in-memory indexer
+  -> C.AddressAny               -- ^ Address to filter for
+  -> IO Result                  -- ^ search results
+queryPlusVolatile ix addr =
+  Ix.getEvents (ix ^. Ix.storage)  >>= query ix addr
```

Comment:
@raduom 
the Index.VSplit has two functions, getEvents and getBuffer. My take is that getBuffer returns the events that are about to `batched` to SQL.  
Where as getEvents, are events that are `Volatile`, they may get rolled back.  
Do I have that correct?

Reply:
Yes.